eth/beacon/state_machines/forks/serenity/epoch_transitions.py

```diff
@@ -0,0 +1,224 @@
+)
+from eth.beacon.typing import (
+    Ether,
+    Gwei,
+    SlotNumber,
+)
+from eth.beacon.types.validator_records import ValidatorRecord
+from eth.beacon.types.states import BeaconState
+from eth.beacon.state_machines.configs import BeaconConfig
+from eth.beacon.helpers import (
+    get_effective_balance,
+    get_active_validator_indices,
+    get_attestation_participants,
+    get_block_root,
+    total_balance,
+)
+from eth.beacon.constants import TWO_POWER_64
+
+
+def get_epoch_boundary_attesting_balances(
+        state: BeaconState,
+        config: BeaconConfig) -> Tuple[Gwei, Gwei]:
+    """
+    previous_epoch_boundary_attesting_balance:
+        SELECT sum(v.balance)
+        FROM Validator as v
+        RIGHT JOIN Attestations as a
+        ON v.index = a.participant
+        WHERE
+            a.slot between `now` and `2 epoch ago` AND
+            a.justified_slot = previous_justified_slot AND
+            a.epoch_boundary_root = `2 epoch ago`
+        GROUP BY v.index
+
+    current_epoch_boundary_attesting_balance:
+        SELECT sum(v.balance)
+        FROM Validator as v
+        RIGHT JOIN Attestations as a
+        ON v.index = a.participant
+        WHERE
+            a.slot between `now` and `1 epoch ago` AND
+            a.justified_slot = justified_slot AND
+            a.epoch_boundary_root = `1 epoch ago`
+        GROUP BY v.index
+    """
+
+    EPOCH_LENGTH = config.EPOCH_LENGTH
+    MAX_DEPOSIT = config.MAX_DEPOSIT
+    LATEST_BLOCK_ROOTS_LENGTH = config.LATEST_BLOCK_ROOTS_LENGTH
+
+    current_epoch_attestations = tuple(
+        attestation
+        for attestation in state.latest_attestations
+        if state.slot - EPOCH_LENGTH <= attestation.data.slot < state.slot
+    )
+    previous_epoch_attestations = tuple(
+        attestation
+        for attestation in state.latest_attestations
+        if state.slot - 2 * EPOCH_LENGTH <= attestation.data.slot < state.slot - EPOCH_LENGTH
+    )
+
+    previous_epoch_justified_attestations = tuple(
+        attestation
+        for attestation in current_epoch_attestations + previous_epoch_attestations
+        if attestation.justified_slot == state.previous_justified_slot
+    )
+
+    previous_epoch_boundary_root = get_block_root(
+        state,
+        state.slot - 2 * EPOCH_LENGTH,
+        LATEST_BLOCK_ROOTS_LENGTH)
+    previous_epoch_boundary_attestations = tuple(
+        attestation
+        for attestation in previous_epoch_justified_attestations
+        if attestation.epoch_boundary_root == previous_epoch_boundary_root
+    )
+
+    sets_of_previous_epoch_boundary_participants = tuple(
+        frozenset(get_attestation_participants(
+            state,
+            attestation.data.slot,
+            attestation.data.shard,
+            attestation.participation_bitfield,
+            EPOCH_LENGTH,
+        ))
+        for attestation in previous_epoch_boundary_attestations
+    )
+    previous_epoch_boundary_attester_indices = (
+        tuple()
```

Comment:
Is there a reason this is a `tuple()` instead of an empty `frozenset()`

Reply:
Thanks, will fix this.